core/src/main/java/io/questdb/cairo/CairoMetadata.java

```diff
@@ -0,0 +1,328 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo;
+
+import io.questdb.cairo.vm.Vm;
+import io.questdb.cairo.vm.api.MemoryCMR;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.*;
+import io.questdb.std.str.*;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.ThreadLocal;
+
+
+public class CairoMetadata implements Sinkable {
+    public static final CairoMetadata INSTANCE = new CairoMetadata();
+    public static final Log LOG = LogFactory.getLog(CairoMetadata.class);
+    private final ConcurrentHashMap<CairoTable> tables = new ConcurrentHashMap<>();
+    ThreadLocal<ColumnVersionReader> tlColumnVersionReader = ThreadLocal.withInitial(ColumnVersionReader::new);
+    ThreadLocal<Path> tlPath = ThreadLocal.withInitial(Path::new);
+
+    public CairoMetadata() {
+    }
+
+    public void clear() {
+        tables.clear();
+    }
+
+    public CairoTable getTableQuiet(@NotNull TableToken tableToken) {
+        return tables.get(tableToken.getDirName());
+    }
+
+    public int getTablesCount() {
+        return tables.size();
+    }
+
+    /**
+     * Hydrates table metadata, bypassing TableWriter/Reader.
+     *
+     * @param token
+     * @param configuration
+     * @param path
+     * @param columnVersionReader
+     * @param blindUpsert         Specifies whether upsert is blind or only if non-null. This is important as TableWriter should take priority over other processes calling this function (async hydration job, queries)
+     */
+    public void hydrateTable(@NotNull TableToken token, @NotNull CairoConfiguration configuration, @NotNull Path path, @NotNull ColumnVersionReader columnVersionReader, boolean blindUpsert, boolean infoLog) {
+        if (infoLog) {
+            LOG.info().$("hydrating metadata [table=").$(token.getTableName()).I$();
+        }
+
+        // set up dir path
+        path.of(configuration.getRoot())
+                .concat(token.getDirName());
+
+        boolean isSoftLink = Files.isSoftLink(path.$());
+
+        // set up table path
+        path.concat(TableUtils.META_FILE_NAME)
+                .trimTo(path.size());
+
+        MemoryCMR metaMem = Vm.getCMRInstance();
+
+        try {
+            // open metadata
+            metaMem.smallFile(configuration.getFilesFacade(), path.$(), MemoryTag.NATIVE_METADATA_READER);
```

Comment:
it looks like there isnt fuzz test around this area. This code has to be highly defensive and not accept on the face value that table has not been deleted